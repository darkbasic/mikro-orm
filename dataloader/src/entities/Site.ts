import {Entity, Property, ManyToMany, Collection, OneToMany} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Sport} from './Sport';
import {Match} from './Match';

interface SiteConstructor {
  name: string;
  position: string;
  sports: Sport[];
}

export {Site as SiteEntity};

@Entity()
export class Site extends BaseEntity {
  @Property()
  name: string;

  @Property()
  position: string;

  @ManyToMany()
  sports: Collection<Sport> = new Collection<Sport>(this);

  @OneToMany(
    () => Match,
    match => match.site
  )
  matches: Collection<Match> = new Collection<Match>(this);

  constructor({name, position, sports}: SiteConstructor) {
    super();
    this.name = name;
    this.position = position;
    this.sports.add(...sports);
  }
}
