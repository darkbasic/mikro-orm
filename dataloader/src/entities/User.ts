import {
  Entity,
  Property,
  Enum,
  OneToMany,
  Collection,
  ManyToMany,
} from 'mikro-orm';
import {BaseEntity} from './BaseEntity';
import {Match} from './Match';

export enum Sex {
  MALE,
  FEMALE,
}

interface UserContructor {
  id?: number;
  email: string;
  name: string;
  surname: string;
  sex: Sex;
  password: string;
  picture?: string;
}

export {User as UserEntity};

@Entity()
export class User extends BaseEntity {
  @Property()
  email: string;

  @Property()
  name: string;

  @Property()
  surname: string;

  @Enum()
  sex: Sex;

  @Property()
  password: string;

  @Property()
  picture?: string;

  @OneToMany(() => Match, match => match.creator)
  createdMatches: Collection<Match> = new Collection<Match>(this);

  @ManyToMany(() => Match, match => match.partecipants)
  partecipatedMatches: Collection<Match> = new Collection<Match>(this);

  constructor({
    id,
    email,
    name,
    surname,
    sex,
    password,
    picture,
  }: UserContructor) {
    super({id});
    this.email = email;
    this.name = name;
    this.surname = surname;
    this.sex = sex;
    this.password = password;
    if (picture) {
      this.picture = picture;
    }
  }
}
