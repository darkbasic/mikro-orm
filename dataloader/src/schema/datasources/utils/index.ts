import {GraphQLResolveInfo} from 'graphql';
import {
  Reference,
  AnyEntity,
  EntityManager,
  Primary,
  IdentifiedReference,
  Collection,
  WrappedEntity,
  wrap,
  EntityRepository,
} from 'mikro-orm';
import DataLoader from 'dataloader';

export function ensureIsNum(id: string | number): number;
export function ensureIsNum(
  id: string | number | null | undefined
): number | undefined;
export function ensureIsNum(ids: Array<string | number>): number[];
export function ensureIsNum(
  ids: Array<string | number> | null | undefined
): number[] | undefined;
export function ensureIsNum(
  idOrIds: null | undefined | string | number | (string | number)[]
): undefined | number | number[] {
  if (idOrIds === null || idOrIds === undefined) {
    return undefined;
  }
  if (idOrIds instanceof Array) {
    return idOrIds.map(id => {
      const num = Number(id);
      if (isNaN(num)) {
        throw new Error(`The ID ${id} is not a valid number`);
      }
      return num;
    });
  } else {
    const num = Number(idOrIds);
    if (isNaN(num)) {
      throw new Error(`The ID ${idOrIds} is not a valid number`);
    }
    return num;
  }
}

export function shouldPopulate({fieldNodes}: GraphQLResolveInfo): boolean {
  if (fieldNodes.length === 1) {
    const selections = fieldNodes[0].selectionSet?.selections;
    if (selections?.length === 1) {
      if (selections[0].kind !== 'InlineFragment') {
        return selections[0].name.value !== 'id';
      }
    }
  }
  return true;
}

export function getItemNoPopulate<T>(
  ref: Reference<T>,
  info?: GraphQLResolveInfo
): T | undefined {
  return !info || shouldPopulate(info) ? undefined : ((ref as unknown) as T);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function getAllPropertyNames(obj: any) {
  const props: string[] = [];

  do {
    Object.getOwnPropertyNames(obj).forEach(prop => {
      if (props.indexOf(prop) === -1) {
        props.push(prop);
      }
    });
  } while ((obj = Object.getPrototypeOf(obj)));

  return props;
}

interface Id {
  id: number | string;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function hasRef<T>(
  entity: T
): T &
  WrappedEntity<T, keyof T> &
  Record<string, IdentifiedReference<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, IdentifiedReference<AnyEntity>>;
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function hasCol<T>(
  entity: T
): T & WrappedEntity<T, keyof T> & Record<string, Collection<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, Collection<AnyEntity>>;
}

function hasRefOrCol<T>(
  entity: T
): T &
  WrappedEntity<T, keyof T> &
  Record<string, IdentifiedReference<AnyEntity> | Collection<AnyEntity>> {
  return entity as T &
    WrappedEntity<T, keyof T> &
    Record<string, IdentifiedReference<AnyEntity> | Collection<AnyEntity>>;
}

function isRef<T>(
  refOrCol: IdentifiedReference<T> | Collection<T>
): refOrCol is IdentifiedReference<T> {
  return !(refOrCol instanceof Collection);
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function isCol<T>(
  refOrCol: IdentifiedReference<T> | Collection<T>
): refOrCol is Collection<T> {
  return refOrCol instanceof Collection;
}

type RepoFind<K> = {
  repo: EntityRepository<K>;
  filter: Record<string, Primary<K> | Primary<K>[]>;
  many: boolean;
};

type Result<K> = {
  entityName: string;
  keys: string[];
  entitiesOrError: K[] | Error;
};

function toArray<T>(arrOrAny: T | T[]): T[] {
  return Array.isArray(arrOrAny) ? arrOrAny : [arrOrAny];
}

function objectsHaveSameKeys<T>(
  ...objects: Record<string, Primary<T> | Primary<T>[]>[]
): boolean {
  const allKeys = new Set(
    objects.reduce(
      (keys, object) => keys.concat(Object.keys(object)),
      [] as string[]
    )
  );
  return objects.every(object => allKeys.size === Object.keys(object).length);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function arrayAreEqual<T extends Primary<any>[] | string[]>(
  a: T,
  b: T
): boolean {
  return JSON.stringify(a.sort()) === JSON.stringify(b.sort());
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function arrayIncludes<T extends (string | Primary<any>)[]>(
  arr: T,
  target: T
): boolean {
  return target.every(v => arr.includes(v));
}

export function groupPrimaryKeysByEntity<T extends AnyEntity<T>>(
  refs: readonly Reference<T>[]
) {
  return refs.reduce((acc, cur) => {
    const {className} = cur.__meta;
    const primaryKeys: Primary<T>[] = acc[className];
    const primaryKey: Primary<T> = cur.__primaryKey;
    acc[className] = primaryKeys
      ? [...new Set([...primaryKeys, primaryKey])]
      : [primaryKey];
    return acc;
  }, {} as Record<string, Primary<T>[]>);
}

export function groupInversedOrMappedKeysByEntity<T extends AnyEntity<T>>(
  collections: readonly Collection<T>[]
): Record<string, Record<string, Primary<T>[]>> {
  return collections.reduce((acc, cur) => {
    const className = cur.property.type;
    const record = acc[className] || (acc[className] = {});
    // Many to Many vs One to Many
    const inversedOrMappedBy = cur.property.inversedBy
      ? 'inversedBy'
      : cur.property.mappedBy
      ? 'mappedBy'
      : undefined;
    if (!inversedOrMappedBy) {
      throw new Error('Cannot find inversedBy or mappedBy prop');
    }
    const key: Primary<T> = cur.owner.__primaryKey;
    const keys: Primary<T>[] = record[cur.property[inversedOrMappedBy]];
    record[cur.property[inversedOrMappedBy]] = keys
      ? [...new Set([...keys, key])]
      : [key];
    return acc;
  }, {} as Record<string, Record<string, Primary<T>[]>>);
}

export function groupFindQueries<T extends AnyEntity<T>>(
  repoFinds: readonly RepoFind<T>[]
) {
  return repoFinds.reduce((acc, {repo, filter}) => {
    const entityName = repo['entityName'] as string;
    if (acc[entityName]) {
      let matchFound = false;
      const curFilters = acc[entityName].map(el => ({...el}));
      for (const curFilter of curFilters) {
        if (objectsHaveSameKeys(curFilter, filter)) {
          matchFound = true;
          Object.entries(filter).forEach(([prop, idOrIds]) => {
            const curIdOrIds = curFilter[prop];
            if (
              Array.isArray(curIdOrIds) ||
              Array.isArray(idOrIds) ||
              (curIdOrIds && curIdOrIds !== idOrIds)
            ) {
              curFilter[prop] = [
                ...new Set([...toArray(curIdOrIds ?? []), ...toArray(idOrIds)]),
              ];
            } else {
              curFilter[prop] = idOrIds;
            }
          });
          break;
        }
      }
      if (matchFound) {
        acc[entityName] = [...curFilters];
      } else {
        acc[entityName] = [...curFilters, filter];
      }
    } else {
      acc[entityName] = [filter];
    }
    return acc;
  }, {} as Record<string, Record<string, Primary<T> | Primary<T>[]>[]>);
}

export class EntityDataLoader<T extends Id = Id, K = AnyEntity<T, 'id'>> {
  private bypass: boolean;
  private refLoader: DataLoader<IdentifiedReference<K>, K>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private colLoader: DataLoader<Collection<any>, K[]>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private findLoader: DataLoader<RepoFind<any>, K[] | K | null>;

  constructor(em: EntityManager, bypass = false) {
    this.bypass = bypass;

    this.refLoader = new DataLoader<IdentifiedReference<K>, K>(async refs => {
      const groupedIds = groupPrimaryKeysByEntity(refs);
      const promises = Object.entries(groupedIds).map(([entity, ids]) =>
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        em.getRepository<K>(entity).find(ids as any)
      );
      await Promise.all(promises);
      return Promise.all(refs.map(async ref => await ref.load()));
    });

    this.colLoader = new DataLoader<Collection<K>, K[]>(async collections => {
      const groupedKeys = groupInversedOrMappedKeysByEntity(collections);
      const promises: Promise<K[]>[] = Object.entries(
        groupedKeys
      ).map(([entity, record]) =>
        em.getRepository<K>(entity).find(record, Object.keys(record))
      );
      const results = (await Promise.all(promises))
        .flat()
        .map(entity => wrap(entity));
      return collections.map(collection =>
        results.filter(result => {
          // Class matches
          if (result.__meta.className === collection.property.type) {
            const inversedOrMappedBy = collection.property.inversedBy
              ? 'inversedBy'
              : collection.property.mappedBy
              ? 'mappedBy'
              : undefined;
            if (!inversedOrMappedBy) {
              throw new Error('Cannot find inversedBy or mappedBy prop');
            }
            const refOrCol = hasRefOrCol(result)[
              collection.property[inversedOrMappedBy]
            ];
            return isRef(refOrCol)
              ? refOrCol.__primaryKey === collection.owner.__primaryKey
              : refOrCol
                  .getItems()
                  .map(el => el.__primaryKey)
                  .includes(collection.owner.__primaryKey);
          }
          return false;
        })
      );
    });

    this.findLoader = new DataLoader<RepoFind<K>, K[] | K | null>(
      async repoFinds => {
        const groupedKeys = groupFindQueries(repoFinds);
        const groupedResults: Result<K>[] = await Promise.all(
          Object.entries(groupedKeys)
            .map(([entityName, records]) => {
              return records.map(async record => {
                let entitiesOrError: K[] | Error;
                try {
                  entitiesOrError = await em
                    .getRepository<K>(entityName)
                    .find(record, Object.keys(record));
                } catch (e) {
                  entitiesOrError = e as Error;
                }
                return {
                  entityName,
                  keys: Object.keys(record),
                  entitiesOrError,
                };
              });
            })
            .flat()
        );

        return repoFinds.map(({repo, filter, many}) => {
          const result = groupedResults.find(
            ({entityName, keys, entitiesOrError}) => {
              if (
                entityName === String(repo['entityName']) &&
                arrayAreEqual(Object.keys(filter), keys)
              ) {
                if (!(entitiesOrError instanceof Error)) {
                  return true;
                } else {
                  throw entitiesOrError;
                }
              }
              return false;
            }
          );
          if (!result) {
            throw new Error('Cannot match results');
          }

          const {entitiesOrError} = result;
          if (!(entitiesOrError instanceof Error)) {
            return (
              entitiesOrError[many ? 'filter' : 'find'](entity => {
                for (const key of Object.keys(filter)) {
                  if (
                    !arrayIncludes(
                      toArray(filter[key]),
                      // eslint-disable-next-line @typescript-eslint/no-explicit-any
                      toArray((entity as any)[key]).map(
                        el => el.id as Primary<K>
                      )
                    )
                  ) {
                    return false;
                  }
                }
                return true;
              }) ?? null
            );
          } else {
            throw entitiesOrError;
          }
        });
      }
    );
  }

  load<L extends K>(ref: IdentifiedReference<L>, bypass?: boolean): Promise<L>;
  load<L extends K>(collection: Collection<L>, bypass?: boolean): Promise<L[]>;
  load<L extends K>(
    refOrCol: IdentifiedReference<L> | Collection<L>,
    bypass?: boolean
  ): Promise<L> | Promise<L[]> {
    if (isRef(refOrCol)) {
      return bypass ?? this.bypass
        ? refOrCol.load()
        : (this.refLoader.load(refOrCol) as Promise<L>);
    } else {
      return bypass ?? this.bypass
        ? refOrCol.loadItems()
        : (this.colLoader.load(refOrCol) as Promise<L[]>);
    }
  }

  find<L extends K>(
    repo: EntityRepository<L>,
    filter: Record<string, Primary<L> | Primary<L>[]>,
    bypass?: boolean
  ): Promise<L[]> {
    return bypass ?? this.bypass
      ? repo.find(filter)
      : (this.findLoader.load({repo, filter, many: true}) as Promise<L[]>);
  }

  findOne<L extends K>(
    repo: EntityRepository<L>,
    filter: Record<string, Primary<L> | Primary<L>[]>,
    bypass?: boolean
  ): Promise<L | null> {
    return bypass ?? this.bypass
      ? repo.findOne(filter)
      : (this.findLoader.load({
          repo,
          filter,
          many: false,
        }) as Promise<L | null>);
  }
}
