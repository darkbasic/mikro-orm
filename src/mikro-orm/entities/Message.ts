import {
  Entity,
  Property,
  ManyToOne,
  IdentifiedReference,
  Reference,
} from '@mikro-orm/core';
import {BaseIdEntity, BaseIdEntityConstructor} from './BaseIdEntity';
import {User} from './User';

interface MessageConstructor extends BaseIdEntityConstructor {
  content: string;
  sender: User;
  recipient: User;
}

@Entity()
export class Message extends BaseIdEntity {
  @Property()
  content: string;

  @ManyToOne()
  sender: IdentifiedReference<User>;

  @ManyToOne()
  recipient: IdentifiedReference<User>;

  constructor({id, createdAt, content, sender, recipient}: MessageConstructor) {
    super({id, createdAt});
    this.content = content;
    this.sender = Reference.create(sender);
    this.recipient = Reference.create(recipient);
  }
}
