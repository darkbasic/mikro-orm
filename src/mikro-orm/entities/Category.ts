import {Entity, Property, OneToMany, Collection} from '@mikro-orm/core';
import {BaseIdEntity, BaseIdEntityConstructor} from './BaseIdEntity';
import {Event} from './Event';

interface CategoryConstructor extends BaseIdEntityConstructor {
  name: string;
}

@Entity()
export class Category extends BaseIdEntity {
  @Property()
  name: string;

  @OneToMany(() => Event, event => event.category)
  events: Collection<Event> = new Collection<Event>(this);

  constructor({id, createdAt, name}: CategoryConstructor) {
    super({id, createdAt});
    this.name = name;
  }
}
