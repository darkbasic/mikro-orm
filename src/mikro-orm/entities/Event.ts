import {
  Entity,
  ManyToMany,
  Collection,
  ManyToOne,
  IdentifiedReference,
  Reference,
  Property,
} from '@mikro-orm/core';
import {BaseIdEntity, BaseIdEntityConstructor} from './BaseIdEntity';
import {User} from './User';
import {Category} from './Category';
import {Site} from './Site';

interface EventConstructor extends BaseIdEntityConstructor {
  name: string;
  creator: User;
  category: Category;
  site: Site;
  partecipants: User[];
}

@Entity()
export class Event extends BaseIdEntity {
  @Property()
  name: string;

  @ManyToOne()
  creator: IdentifiedReference<User>;

  @ManyToOne()
  category: IdentifiedReference<Category>;

  @ManyToOne()
  site: IdentifiedReference<Site>;

  @ManyToMany()
  partecipants: Collection<User> = new Collection<User>(this);

  constructor({
    id,
    name,
    createdAt,
    creator,
    category,
    site,
    partecipants,
  }: EventConstructor) {
    super({id, createdAt});
    this.name = name;
    this.creator = Reference.create(creator);
    this.category = Reference.create(category);
    this.site = Reference.create(site);
    this.partecipants.add(...partecipants);
  }
}
