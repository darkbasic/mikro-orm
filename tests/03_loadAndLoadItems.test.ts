import {MikroORM} from '@mikro-orm/core';
import {PostgreSqlDriver} from '@mikro-orm/postgresql';
import {initORMPostgreSql, resetDatabase, sleep, Optional} from './utils';
import {User} from '../src/mikro-orm/entities/User';
import {Event} from '../src/mikro-orm/entities/Event';
import {EntityDataLoader} from '../src/dataloader/EntityDataLoader';
import {Category} from '../src/mikro-orm/entities/Category';
import {Message} from '../src/mikro-orm/entities/Message';
import {Site} from '../src/mikro-orm/entities/Site';

describe('Load and loadItems', () => {
  jest.setTimeout(10e3);
  let orm: MikroORM<PostgreSqlDriver>;
  let dataloader: EntityDataLoader;

  beforeAll(async () => {
    orm = await initORMPostgreSql();
    await resetDatabase(orm);
  });

  beforeEach(async () => {
    orm.em.clear();
    dataloader = new EntityDataLoader(orm.em);
    await sleep(100);
  });

  const tenEventIds = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19];

  it('should be able to deal with different entities/references/collections in the same event loop', async () => {
    let events = await orm.em.find(Event, {id: tenEventIds});
    let messages = await orm.em.find(Message, {sender: 1});
    const partecipants = await Promise.all(
      events.map(event => event.partecipants.loadItems())
    );
    const categories = await Promise.all(
      events.map(event => event.category.load())
    );
    const sites = await Promise.all(events.map(event => event.site.load()));
    const senders = await Promise.all(
      messages.map(message => message.sender.load())
    );
    const recipients = await Promise.all(
      messages.map(message => message.recipient.load())
    );
    const partecipantsJSON = JSON.stringify(partecipants);
    const categoriesJSON = JSON.stringify(categories);
    const sitesJSON = JSON.stringify(sites);
    const sendersJSON = JSON.stringify(senders);
    const recipientsJSON = JSON.stringify(recipients);
    expect(partecipants).toBeDefined();
    expect(partecipants.length).toBeGreaterThan(0);
    expect(partecipants).toMatchSnapshot();
    expect(categories).toBeDefined();
    expect(categories.length).toBeGreaterThan(0);
    expect(categories).toMatchSnapshot();
    expect(sites).toBeDefined();
    expect(sites.length).toBeGreaterThan(0);
    expect(sites).toMatchSnapshot();
    expect(senders).toBeDefined();
    expect(senders.length).toBeGreaterThan(0);
    expect(senders).toMatchSnapshot();
    expect(recipients).toBeDefined();
    expect(recipients.length).toBeGreaterThan(0);
    expect(recipients).toMatchSnapshot();

    orm.em.clear();

    events = await orm.em.find(Event, {id: tenEventIds});
    messages = await orm.em.find(Message, {sender: 1});
    const resDataloader = await Promise.all<User[] | User | Category | Site>([
      ...events.map(event => dataloader.load(event.partecipants)),
      ...events.map(event => dataloader.load(event.category)),
      ...events.map(event => dataloader.load(event.site)),
      ...messages.map(message => dataloader.load(message.sender)),
      ...messages.map(message => dataloader.load(message.recipient)),
    ]);
    let [i, j] = [0, partecipants.length];
    const partecipantsDataloader = resDataloader.slice(i, j) as Optional<
      User,
      'partecipatedEvents'
    >[][];
    partecipantsDataloader.forEach(users => {
      users.sort((a, b) => a.id - b.id);
      users.forEach(el => delete el.partecipatedEvents);
    });
    expect(partecipantsDataloader).toBeDefined();
    expect(partecipantsDataloader.length).toBe(partecipants.length);
    expect(JSON.stringify(partecipantsDataloader)).toBe(partecipantsJSON);
    [i, j] = [i + partecipants.length, j + categories.length];
    const categoriesDataloader = resDataloader.slice(i, j) as Category[];
    expect(categoriesDataloader).toBeDefined();
    expect(categoriesDataloader.length).toBe(categories.length);
    expect(JSON.stringify(categoriesDataloader)).toBe(categoriesJSON);
    [i, j] = [i + categories.length, j + sites.length];
    const sitesDataloader = resDataloader.slice(i, j) as Site[];
    expect(sitesDataloader).toBeDefined();
    expect(sitesDataloader.length).toBe(sites.length);
    expect(JSON.stringify(sitesDataloader)).toBe(sitesJSON);
    [i, j] = [i + sites.length, j + senders.length];
    const sendersDataloader = resDataloader.slice(i, j) as User[];
    expect(sendersDataloader).toBeDefined();
    expect(sendersDataloader.length).toBe(senders.length);
    expect(JSON.stringify(sendersDataloader)).toBe(sendersJSON);
    [i, j] = [i + senders.length, j + recipients.length];
    const recipientsDataloader = resDataloader.slice(i, j) as User[];
    expect(recipientsDataloader).toBeDefined();
    expect(recipientsDataloader.length).toBe(recipients.length);
    expect(JSON.stringify(recipientsDataloader)).toBe(recipientsJSON);
  });

  it('should be faster with EntityDataLoader', async () => {
    let events = await orm.em.find(Event, {id: tenEventIds});
    let messages = await orm.em.find(Message, {sender: 1});
    const start = new Date();
    await Promise.all<User[] | User | Category | Site>([
      ...events.map(event => event.partecipants.loadItems()),
      ...events.map(event => event.category.load()),
      ...events.map(event => event.site.load()),
      ...messages.map(message => message.sender.load()),
      ...messages.map(message => message.recipient.load()),
    ]);
    const end = new Date();
    const diff = end.getTime() - start.getTime();

    orm.em.clear();

    events = await orm.em.find(Event, {id: tenEventIds});
    messages = await orm.em.find(Message, {sender: 1});
    const startDataloader = new Date();
    await Promise.all<User[] | User | Category | Site>([
      ...events.map(event => dataloader.load(event.partecipants)),
      ...events.map(event => dataloader.load(event.category)),
      ...events.map(event => dataloader.load(event.site)),
      ...messages.map(message => dataloader.load(message.sender)),
      ...messages.map(message => dataloader.load(message.recipient)),
    ]);
    const endDataloader = new Date();
    const diffDataloader = endDataloader.getTime() - startDataloader.getTime();
    console.log(`W/o dataloader: ${diff} ms`);
    console.log(`W/ dataloader: ${diffDataloader} ms`);
    expect(diffDataloader).toBeLessThanOrEqual(diff);
  });

  afterEach(async () => {});

  afterAll(async () => {
    orm.close(true);
    await sleep(100);
  });
});
