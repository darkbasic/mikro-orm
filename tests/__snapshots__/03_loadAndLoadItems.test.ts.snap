// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Load and loadItems should be able to deal with different entities/references/collections in the same event loop 1`] = `
Array [
  Array [
    Object {
      "createdAt": 2030-02-10T14:45:00.000Z,
      "id": 1,
      "name": "Niccolò",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-02-11T09:10:00.000Z,
      "id": 2,
      "name": "Mario",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-01T19:30:00.000Z,
      "id": 4,
      "name": "Mattia",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-05T15:20:00.000Z,
      "id": 6,
      "name": "Filippo",
      "sex": 0,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-10T14:45:00.000Z,
      "id": 1,
      "name": "Niccolò",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-02-11T09:10:00.000Z,
      "id": 2,
      "name": "Mario",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-01T19:30:00.000Z,
      "id": 4,
      "name": "Mattia",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-05T15:20:00.000Z,
      "id": 6,
      "name": "Filippo",
      "sex": 0,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-03-06T17:50:00.000Z,
      "id": 7,
      "name": "Sara",
      "sex": 1,
    },
    Object {
      "createdAt": 2030-03-06T20:30:00.000Z,
      "id": 8,
      "name": "Lucia",
      "sex": 1,
    },
    Object {
      "createdAt": 2030-03-06T21:50:00.000Z,
      "id": 9,
      "name": "Giorgia",
      "sex": 1,
    },
    Object {
      "createdAt": 2030-03-07T07:30:00.000Z,
      "id": 10,
      "name": "Claudia",
      "sex": 1,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-11T09:10:00.000Z,
      "id": 2,
      "name": "Mario",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-01T19:30:00.000Z,
      "id": 4,
      "name": "Mattia",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-05T15:20:00.000Z,
      "id": 6,
      "name": "Filippo",
      "sex": 0,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-10T14:45:00.000Z,
      "id": 1,
      "name": "Niccolò",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-02-11T09:10:00.000Z,
      "id": 2,
      "name": "Mario",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-01T19:30:00.000Z,
      "id": 4,
      "name": "Mattia",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-05T15:20:00.000Z,
      "id": 6,
      "name": "Filippo",
      "sex": 0,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-10T14:45:00.000Z,
      "id": 1,
      "name": "Niccolò",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-02-11T09:10:00.000Z,
      "id": 2,
      "name": "Mario",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-01T19:30:00.000Z,
      "id": 4,
      "name": "Mattia",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-05T15:20:00.000Z,
      "id": 6,
      "name": "Filippo",
      "sex": 0,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-10T14:45:00.000Z,
      "id": 1,
      "name": "Niccolò",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-02-11T09:10:00.000Z,
      "id": 2,
      "name": "Mario",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-01T19:30:00.000Z,
      "id": 4,
      "name": "Mattia",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-05T15:20:00.000Z,
      "id": 6,
      "name": "Filippo",
      "sex": 0,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-10T14:45:00.000Z,
      "id": 1,
      "name": "Niccolò",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-02-15T08:15:00.000Z,
      "id": 3,
      "name": "Luigi",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-06T20:30:00.000Z,
      "id": 8,
      "name": "Lucia",
      "sex": 1,
    },
    Object {
      "createdAt": 2030-03-07T07:30:00.000Z,
      "id": 10,
      "name": "Claudia",
      "sex": 1,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-02-15T08:15:00.000Z,
      "id": 3,
      "name": "Luigi",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-02T18:00:00.000Z,
      "id": 5,
      "name": "Riccardo",
      "sex": 0,
    },
    Object {
      "createdAt": 2030-03-06T20:30:00.000Z,
      "id": 8,
      "name": "Lucia",
      "sex": 1,
    },
  ],
  Array [
    Object {
      "createdAt": 2030-03-06T20:30:00.000Z,
      "id": 8,
      "name": "Lucia",
      "sex": 1,
    },
    Object {
      "createdAt": 2030-03-07T07:30:00.000Z,
      "id": 10,
      "name": "Claudia",
      "sex": 1,
    },
  ],
]
`;

exports[`Load and loadItems should be able to deal with different entities/references/collections in the same event loop 2`] = `
Array [
  Object {
    "createdAt": 2030-02-01T00:01:00.000Z,
    "id": 1,
    "name": "Concert",
  },
  Object {
    "createdAt": 2030-02-01T00:01:00.000Z,
    "id": 1,
    "name": "Concert",
  },
  Object {
    "createdAt": 2030-03-02T01:02:00.000Z,
    "id": 2,
    "name": "Play",
  },
  Object {
    "createdAt": 2030-02-01T00:01:00.000Z,
    "id": 1,
    "name": "Concert",
  },
  Object {
    "createdAt": 2030-02-01T00:01:00.000Z,
    "id": 1,
    "name": "Concert",
  },
  Object {
    "createdAt": 2030-02-01T00:01:00.000Z,
    "id": 1,
    "name": "Concert",
  },
  Object {
    "createdAt": 2030-02-01T00:01:00.000Z,
    "id": 1,
    "name": "Concert",
  },
  Object {
    "createdAt": 2030-04-03T01:03:00.000Z,
    "id": 3,
    "name": "Ballad",
  },
  Object {
    "createdAt": 2030-04-03T01:03:00.000Z,
    "id": 3,
    "name": "Ballad",
  },
  Object {
    "createdAt": 2030-04-03T01:03:00.000Z,
    "id": 3,
    "name": "Ballad",
  },
]
`;

exports[`Load and loadItems should be able to deal with different entities/references/collections in the same event loop 3`] = `
Array [
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 1,
    "name": "The Beach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 2,
    "name": "Palabeach",
  },
  Object {
    "createdAt": 2030-02-05T07:15:00.000Z,
    "id": 2,
    "name": "Palabeach",
  },
]
`;

exports[`Load and loadItems should be able to deal with different entities/references/collections in the same event loop 4`] = `
Array [
  Object {
    "createdAt": 2030-02-10T14:45:00.000Z,
    "id": 1,
    "name": "Niccolò",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-02-10T14:45:00.000Z,
    "id": 1,
    "name": "Niccolò",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-02-10T14:45:00.000Z,
    "id": 1,
    "name": "Niccolò",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-02-10T14:45:00.000Z,
    "id": 1,
    "name": "Niccolò",
    "sex": 0,
  },
]
`;

exports[`Load and loadItems should be able to deal with different entities/references/collections in the same event loop 5`] = `
Array [
  Object {
    "createdAt": 2030-02-15T08:15:00.000Z,
    "id": 3,
    "name": "Luigi",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-03-01T19:30:00.000Z,
    "id": 4,
    "name": "Mattia",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-03-02T18:00:00.000Z,
    "id": 5,
    "name": "Riccardo",
    "sex": 0,
  },
  Object {
    "createdAt": 2030-03-02T18:00:00.000Z,
    "id": 5,
    "name": "Riccardo",
    "sex": 0,
  },
]
`;
